vcd/resource_vcd_nsxt_edgegateway_dhcp_forwarding.go

```diff
@@ -0,0 +1,213 @@
+}
+
+func resourceVcdNsxtEdgegatewayDhcpForwardingUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	return resourceVcdNsxtEdgegatewayDhcpForwardingCreateUpdate(ctx, d, meta, "update")
+}
+
+func resourceVcdNsxtEdgegatewayDhcpForwardingCreateUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}, method string) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	unlock, err := vcdClient.lockParentVdcGroupOrEdgeGateway(d)
+	if err != nil {
+		return diag.Errorf("[DHCP forwarding %s] %s", method, err)
+	}
+
+	defer unlock()
+
+	orgName := d.Get("org").(string)
+	edgeGatewayId := d.Get("edge_gateway_id").(string)
+
+	nsxtEdge, err := vcdClient.GetNsxtEdgeGatewayById(orgName, edgeGatewayId)
+	if err != nil {
+		return diag.Errorf("[DHCP forwarding %s] error retrieving Edge Gateway: %s", method, err)
+	}
+
+	dhcpForwardingConfig := &types.NsxtEdgeGatewayDhcpForwarder{
+		Enabled:     d.Get("enabled").(bool),
+		DhcpServers: convertSchemaSetToSliceOfStrings(d.Get("dhcp_servers").(*schema.Set)),
+	}
+
+	_, err = nsxtEdge.UpdateDhcpForwarder(dhcpForwardingConfig)
+	if err != nil {
+		return diag.Errorf("[DHCP forwarding %s] error updating DHCP forwarding configuration: %s", method, err)
+	}
+
+	d.SetId(edgeGatewayId)
+
+	return resourceVcdNsxtEdgegatewayDhcpForwardingRead(ctx, d, meta)
+}
+
+func resourceVcdNsxtEdgegatewayDhcpForwardingRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	return genericVcdNsxtEdgegatewayDhcpForwardingRead(ctx, d, meta, "resource")
+}
+
+func genericVcdNsxtEdgegatewayDhcpForwardingRead(_ context.Context, d *schema.ResourceData, meta interface{}, origin string) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	orgName := d.Get("org").(string)
+	edgeGatewayId := d.Get("edge_gateway_id").(string)
+
+	nsxtEdge, err := vcdClient.GetNsxtEdgeGatewayById(orgName, edgeGatewayId)
+	if govcd.ContainsNotFound(err) {
+		// When parent Edge Gateway is not found - this resource is also not found and should be
+		// removed from state
+		if origin == "datasource" {
+			return diag.Errorf("[DHCP forwarding DS read] error retrieving NSX-T Edge Gateway DHCP forwarding: %s", err)
+		}
+		d.SetId("")
+		log.Printf("[DEBUG] Edge gateway no longer exists. Removing from tfstate")
+		return nil
+	}
+
+	if err != nil {
+		return diag.Errorf("[DHCP forwarding read] error: %s", err)
+	}
+
+	dhcpForwardConfig, err := nsxtEdge.GetDhcpForwarder()
+	if err != nil {
+		return diag.Errorf("[DHCP forwarding read] error retrieving NSX-T Edge Gateway DHCP forwarding: %s", err)
+	}
+
+	// DHCP forwarding does not have its own ID - it is a part of Edge Gateway
+	d.SetId(edgeGatewayId)
+	dSet(d, "enabled", dhcpForwardConfig.Enabled)
+
+	err = d.Set("dhcp_servers", convertStringsToTypeSet(dhcpForwardConfig.DhcpServers))
+	if err != nil {
+		return diag.Errorf("error setting dhcp_servers attribute: %s", err)
+	}
+
+	if !dhcpForwardConfig.Enabled {
```

Comment:
This warning is great, but we could make it more precise:
* Use it in "Update" function, not in "read"
* Only trigger it if `!d.get("enabled").(bool) && d.HasChange("dhcp_servers")`

Reply:
Updated, also to handle it in function, it needed some changes, mostly with declaring `diag.Diagnostics` variable and then returning a slice of the variable and the `Read` function's return value.